src/lib.rs

```diff
@@ -16,21 +27,67 @@ const ANY_IN_ERROR_MSG: &str = "Resource misses at least one environment variabl
 const ANY_NOT_IN_ERROR_MSG: &str =
     "Resource should not contain any environment variable from the rule.";
 
+const EXTRACT_LABEL_ANNOTATION_REGEX: &str = r"\[.*\]";
+
 #[no_mangle]
 pub extern "C" fn wapc_init() {
     register_function("validate", validate);
     register_function("validate_settings", validate_settings::<settings::Settings>);
     register_function("protocol_version", protocol_version_guest);
 }
 
-fn validate_envvar_with_rule(rule: &Rule, env_vars: &Vec<apicore::EnvVar>) -> Result<()> {
+fn validate_envvar_with_rule(
+    rule: &Rule,
+    pod: &apicore::PodSpec,
+    metadata: &ObjectMeta,
+    env_vars: &Vec<apicore::EnvVar>,
+) -> Result<()> {
     let error_message: &str;
     let mut resource_env_var: HashSet<settings::EnvVar> = HashSet::new();
+    let regex = Regex::new(EXTRACT_LABEL_ANNOTATION_REGEX).unwrap();
     for envvar in env_vars {
-        resource_env_var.insert(settings::EnvVar {
-            name: envvar.name.clone(),
-            value: envvar.value.clone(),
-        });
+        if let Some(env_var_source) = &envvar.value_from {
+            if let Some(field_ref) = &env_var_source.field_ref {
+                let mut value = None;
+                let field_path = field_ref.field_path.as_str();
+                if field_path == "metadata.name" {
+                    value = Some(metadata.name.clone().unwrap());
+                } else if field_path == "metadata.namespace" {
+                    value = Some(metadata.namespace.clone().unwrap());
+                } else if field_path.starts_with("metadata.labels") {
+                    if let Some(labels) = &metadata.labels {
+                        let cap = regex.captures(field_path).unwrap();
+                        let label_len = &cap[0].len();
+                        let label = &cap[0][2..label_len - 2];
+                        value = labels.get(label).cloned();
+                    } else {
+                        value = None;
+                    }
+                } else if field_path.starts_with("metadata.annotations") {
+                    if let Some(annotations) = &metadata.annotations {
+                        let cap = regex.captures(field_path).unwrap();
+                        let annotation_len = &cap[0].len();
+                        let annotation = &cap[0][2..annotation_len - 2];
+                        value = annotations.get(annotation).cloned();
+                    } else {
+                        value = None;
+                    }
+                } else if field_path == "spec.nodeName" {
+                    value = pod.node_name.clone();
+                } else if field_path == "spec.serviceAccountName" {
+                    value = pod.service_account_name.clone();
+                }
```

Comment:
I would change that into a `match` statement, I find it cleaner

Reply:
I agree. But than I will mixing `if` with `match`. I need to use `if` to be able to do a partial string matching. I just wanna to avoid missing two different structures to do the same thing.